tck/mp-jwt-embedded/src/main/java/org/apache/tomee/microprofile/jwt/cdi/ClaimBean.java

```diff
@@ -0,0 +1,361 @@
+                            return Optional.ofNullable(claimValue);
+                        });
+
+                    } else if (claimValueType instanceof ParameterizedType && isSet((ParameterizedType) claimValueType)) {
+                        claimValueWrapper.setValue(() -> {
+                            final T claimValue = getClaimValue(key);
+                            return claimValue;
+                        });
+
+                    } else if (claimValueType instanceof ParameterizedType && isList((ParameterizedType) claimValueType)) {
+                        claimValueWrapper.setValue(() -> {
+                            final T claimValue = getClaimValue(key);
+                            return claimValue;
+                        });
+
+                    } else if (claimValueType instanceof Class) {
+                        claimValueWrapper.setValue(() -> {
+                            final T claimValue = getClaimValue(key);
+                            return claimValue;
+                        });
+
+                    } else {
+                        throw new IllegalArgumentException("Unsupported ClaimValue type " + claimValueType.toString());
+                    }
+
+                    return (T) claimValueWrapper;
+                }
+
+                // handle Optional<T>
+                if (rawTypeClass.isAssignableFrom(Optional.class)) {
+                    return getClaimValue(key);
+                }
+
+                // handle Set<T>
+                if (rawTypeClass.isAssignableFrom(Set.class)) {
+                    return getClaimValue(key);
+                }
+
+                // handle List<T>
+                if (rawTypeClass.isAssignableFrom(List.class)) {
+                    return getClaimValue(key);
+                }
+            }
+
+        } else if (annotated.getBaseType().getTypeName().startsWith("javax.json.Json")) {
+            // handle JsonValue<T> (number, string, etc)
+            return (T) toJson(key);
+
+        } else {
+            // handle Raw types
+            return getClaimValue(key);
+        }
+
+        throw new IllegalStateException("Unhandled Claim type " + annotated.getBaseType());
+    }
+
+    public static String getClaimKey(final Claim claim) {
+        return claim.standard() == Claims.UNKNOWN ? claim.value() : claim.standard().name();
+    }
+
+    private T getClaimValue(final String name) {
+        final Bean<?> bean = bm.resolve(bm.getBeans(JsonWebToken.class));
+        if (RequestScoped.class.equals(bean.getScope())) {
+            jsonWebToken = JsonWebToken.class.cast(bm.getReference(bean, JsonWebToken.class, null));
+        }
+        if (jsonWebToken == null || !bean.getScope().equals(RequestScoped.class)) {
+            logger.warning(String.format("Can't retrieve claim %s. No active principal.", name));
+            return null;
+        }
+
+        final Optional<T> claimValue = jsonWebToken.claim(name);
```

Comment:
Definitely Java 8?

Reply:
Yes, hence the reason why we need to extract most of this code into a separate library.